src/Tools/dotnet-monitor/Egress/S3/MultiPartUploadStream.cs

```diff
@@ -0,0 +1,140 @@
+using Amazon.S3;
+using Amazon.S3.Model;
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.Egress.S3;
+
+internal class MultiPartUploadStream : Stream
+{
+    private readonly byte[] _buffer;
+    private int _offset;
+    private readonly string _bucketName;
+    private readonly string _objectKey;
+    private readonly string _uploadId;
+    private readonly IAmazonS3 _client;
+    private readonly List<PartETag> _parts = new();
+    public List<PartETag> Parts => _parts.ToList();
+    public bool Closed { get; private set; }
+    private int _position;
+    public const int MinimumSize = 5 * 1024 * 1024; // the minimum size of an upload part (except for the last part)
+    private readonly int _bufferSize;
+
+    public MultiPartUploadStream(IAmazonS3 client, string bucketName, string objectKey, string uploadId, int bufferSize)
+    {
+        _bufferSize = Math.Max(bufferSize, MinimumSize); // has to be at least the minimum
+        _buffer = ArrayPool<byte>.Shared.Rent(_bufferSize);
+        _offset = 0;
+        _client = client;
+        _bucketName = bucketName;
+        _objectKey = objectKey;
+        _uploadId = uploadId;
+    }
+
+    public override async Task FlushAsync(CancellationToken cancellationToken)
+    {
+        if (Closed)
+            throw new ObjectDisposedException(nameof(MultiPartUploadStream));
+        if (_offset == 0 || _offset < MinimumSize)
```

Comment:
Consider adding a `bool allowPartialWrite = false` parameter to `DoWriteAsync`, check `if (!allowPartialWrite && _offset < MinimuSize)` condition within `DoWriteAsync`, and the just call it from here. Update `FinalizeAsync` to call `DoWriteAsync` with `allowPartialWrite: true`. This will alleviate this method from understanding the writing condition.